projects/hip/docs/understand/hardware_implementation.rst

```diff
@@ -1,167 +1,604 @@
-HIP, and how the :ref:`inherent_thread_model` maps to the hardware.
+This chapter describes the hardware architecture of AMD GPUs supported by HIP,
+focusing on the internal organization and operation of GPU hardware components.
+Understanding these hardware details helps you optimize GPU applications and
+achieve maximum performance. For the programming model that maps to this
+hardware, see :ref:`programming_model`.
 
-Compute units
-=============
+Overall GPU architecture
+========================
+
+AMD GPUs consist of interconnected blocks of digital circuits that work together
+to execute complex parallel computing tasks. The architecture is organized
+hierarchically to enable massive parallelism while managing resources efficiently.
+
+Command processor and control
+-----------------------------
+
+The command processor (CP) serves as the primary interface between the CPU and
+GPU, receiving and distributing commands for execution. The CP consists of two
+main components:
+
+* **Command processor fetcher (CPF)**: Fetches commands from memory and passes
+  them to the CPC for processing.
+* **Command processor packet processor (CPC)**: A microcontroller that decodes
+  fetched commands and dispatches kernels to the workgroup processors for
+  scheduling.
+
+The command processor handles several types of operations:
+
+* Kernel launches, which are forwarded to asynchronous compute engines (ACEs)
+* Memory transfers, which are delegated to direct memory access (DMA) engines
+* Synchronization operations and memory fences
+
+**DMA engines** handle memory transfers between CPU and GPU memory without CPU
+involvement after initialization. Most GPUs contain two DMA engines, enabling
+concurrent bidirectional transfers to better utilize PCIe bandwidth. The DMA
+engines fetch data in small chunks and can process transfers in parallel but
+cannot handle multiple copy commands on the same engine simultaneously.
+
+**Asynchronous compute engines (ACEs)** break down kernels into workgroups for
+distribution to shader processor input (SPI) blocks. Multiple ACEs enable
+concurrent kernel execution, with each ACE capable of dispatching one kernel
+at a time. ACEs process commands from different queues asynchronously, enabling
+overlap between different kernel executions and memory operations.
+
+Hierarchical organization
+-------------------------
+
+The GPU organizes compute resources in a three-level hierarchy that enables
+modular design and resource sharing:
 
-The basic building block of a GPU is a compute unit (CU), also known
-as streaming multiprocessor (SM) on NVIDIA GPUs. The thread blocks making up a
-grid are scheduled for execution on CUs. Each block is assigned to an
-individual CU, and a CU can accommodate several blocks. Depending on their
-resource usage up to thousands of threads can reside on a CU.
+.. figure:: ../data/understand/hardware_implementation/selayout.png
```

Comment:
This figure confuses me as it does not seem to show the "hierarchical organization of compute units grouped into shader engines". Instead it seems to show the CUs and Workgroup processors which are dispatched to by the CPC?